File: 20.valid-parentheses.cpp
/*
 * @lc app=leetcode id=20 lang=cpp
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {
   public:
    bool isValid(string s) {
        stack<char> characters;
        for (char c : s) {
            if (c == '(' || c == '[' || c == '{')
                characters.push(c);
            else if (!characters.empty() &&
                     (c == ')' && characters.top() == '(' ||
                      c == '}' && characters.top() == '{' ||
                      c == ']' && characters.top() == '['))
                characters.pop();
            else
                return false;
        }
        if (characters.empty())
            return true;
        else
            return false;
    }
};
// @lc code=end

File: 20.Valid-Parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  let stack = [];
  let res = true;

  for (let i = 0; i < s.length; ++i) {
    if (s[i] === '(') stack.push('(');
    else if (s[i] === '{') stack.push('{');
    else if (s[i] === '[') stack.push('[');
    else {
      let top = stack.pop();
      if (!(top === '(' && s[i] === ')' || top === '{' && s[i] === '}' || top === '[' && s[i] === ']')) {
        res = false;
        break;
      }
    }
  }

  if (stack.length) res = false;

  return res;
};